cloudinit/config/cc_reset_rmc.py

```diff
@@ -0,0 +1,167 @@
+# (c) Copyright IBM Corp. 2020 All Rights Reserved
+#
+# Author: Aman Kumar Sinha <amansi26@in.ibm.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+
+"""
+Reset RMC
+------------
+**Summary:** reset rsct node id
+
+Reset RMC module is IBM PowerVM Hypervisor specific
+
+Reliable Scalable Cluster Technology (RSCT) is a set of software components,
+that  together provide a comprehensive clustering environment (RAS features)
+for IBM PowerVM based virtual machines. RSCT includes the Resource monitoring
+and control (RMC) subsystem. RMC is a generalized framework used for managing,
+monitoring, and manipulating resources. RMC runs as a daemon process on
+individual machines and needs creation of unique node id and restarts
+during VM boot.
+More details refer
+https://www.ibm.com/support/knowledgecenter/en/SGVKBA_3.2/admin/bl503_ovrv.htm
+
+This module handles
+- creation of the unique RSCT node id to every instance/virtual machine
+  and ensure once set, it isn't changed subsequently by cloud-init.
+  In order to do so, it restarts RSCT service.
+
+Prerequisite of using this module is to install RSCT packages.
+
+**Internal name:** per instance
+
+**Supported distros:** rhel, sles and ubuntu
+
+"""
+
+import json
+import os
+import time
+
+from cloudinit.settings import PER_INSTANCE
+from cloudinit import util
+from cloudinit import subp
+
+frequency = PER_INSTANCE
+PIDOF = 'pidof'
+RMCCTRL = '/usr/sbin/rsct/bin/rmcctrl'
+RECFGCT = '/usr/sbin/rsct/install/bin/recfgct'
+SRCMSTR = '/sbin/srcmstr'
+
+NODE_ID_FILE = '/etc/ct_node_id'
+
+SRCMSTR_TIMEOUT_SECONDS = 30
+
+
+def handle(name, _cfg, _cloud, log, _args):
+    global RMCCTRL, RECFGCT
+    with open('/run/cloud-init/instance-data.json') as data_file:
```

Comment:
I'm not sure how...but we probably hvae some way to get platform here without going through parsing the instance data that cloud-init itself wrote.

Reply:
That part of the content of that file is determined in `cloudinit.sources.DataSource._get_standardized_metadata`, specifically: https://github.com/canonical/cloud-init/blob/5bf287f430b97860bf746e61b83ff53b834592d0/cloudinit/sources/__init__.py#L239 so `ds.platform_type` (where `ds` is an instance of a `DataSource` subclass) should produce the same value.

In a config module, you're passed `cloud` which is a `cloudinit.cloud.Cloud` instance; that will have a `datasource` property which is a `DataSource` subclass. So I believe that (after renaming `_cloud` to `cloud`), `if ds_value == 'none'` should become `if cloud.datasource.platform_type == 'none'`.

All that said, `platform_type` will be the same across reboots of a specific instance (i.e. if it was "ec2" on first boot, it will be "ec2" on all subsequent boots too), so it's not clear to me that that `if` statement is actually guaranteeing that node ID generation only happens on first boot.

Reply:
@Aman306, agree with @OddBloke that if there is a mechanism to get this detail without directly getting it from the file, that should be used. If for some reason that does not work out, then atleast the open() should be replaced by uils.loadfile like @smoser recommended in one of the other comments.